src/components/tl/mlx5/mcast/tl_mlx5_mcast_team.c

```diff
@@ -227,19 +230,50 @@ ucc_status_t ucc_tl_mlx5_mcast_coll_setup_comm_resources(ucc_tl_mlx5_mcast_coll_
     comm->pending_recv = 0;
     comm->buf_n        = comm->params.rx_depth * 2;
 
-    ret = posix_memalign((void**) &comm->pp_buf, page_size, buf_size * comm->buf_n);
-    if (ret) {
-        tl_error(comm->ctx->lib, "posix_memalign failed");
-        return UCC_ERR_NO_MEMORY;
+    supported_mem_type = comm->cuda_mem_enabled ? UCC_MEMORY_TYPE_CUDA
+                                                : UCC_MEMORY_TYPE_HOST;
+
+    status = ucc_mc_alloc(&comm->grh_buf_header, GRH_LENGTH *
+                          sizeof(char), supported_mem_type);
+    comm->grh_buf = comm->grh_buf_header->addr;
+    if (ucc_unlikely(status != UCC_OK)) {
+        tl_error(comm->ctx->lib, "failed to allocate cuda memory");
+        return status;
+    }
+
+    status = ucc_mc_memset(comm->grh_buf, 0, GRH_LENGTH, supported_mem_type);
+    if (status != UCC_OK) {
+        tl_error(comm->ctx->lib, "could not cuda memset");
+        goto error;
+    }
+
+    comm->grh_mr = ibv_reg_mr(comm->ctx->pd, comm->grh_buf, GRH_LENGTH,
+                              IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_LOCAL_WRITE);
+    if (!comm->grh_mr) {
+        tl_error(comm->ctx->lib, "could not register device memory for GRH, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    status = ucc_mc_alloc(&comm->pp_buf_header, buf_size * comm->buf_n, supported_mem_type);
```

Comment:
the only reason that we are using ```ucc_mc_alloc``` is that it gives us a single call that can be used for both CPU and GPU. 
I think this should be fine here, because we are calling ```ucc_mc_free()``` on this memory only when we are freeing the team. I doubt mpool design would modify or touch the memory pool buckets which are in use

Reply:
but definately,  i agree that we should change ```ucc_mc_allocate``` API to have the flexibility to bypass the mpool. In the future we might need it, as mpool is required only when we deal with frequent malloc/free scenarios.

Reply:
I added a ```TODO``` in the code so that we do not forget about this.